### PR TITLE
feat: Hamming distance algorithm.

### PR DESCRIPTION
**Pull request by ravibitsgoa**
#### Description of Change
Added bit manipulation algorithms: hamming distance, power of 2, swapping numbers using xor.

#### Checklist

- [x] Added description of the change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, the tests must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1376"></a>

### CODE REVIEW

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
It is recommended to make one PR per file.
This will make the review process much more faster.

**Comment by ravibitsgoa** [comment]
> It is recommended to make one PR per file.
> This will make the review process much more faster.

Done.

File: bit_manipulation/hamming_distance.cpp [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+    return bitCount(a ^ b);
+}
+
+int main() {
+    int a = 11;  // 1011 in binary
+    int b = 2;   // 0010 in binary
+
+    std::cout << "Hamming distance between " << a << " and " << b << " is "
+         << hamming_distance(a, b) << std::endl;
+}
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:40]:
```suggestion
}

```

File: bit_manipulation/hamming_distance.cpp [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+unsigned int hamming_distance(int a, int b) {
+    if (a < 0 || b < 0) {
+        throw "Both arguments must be >=0 for finding hamming distance.";
+    }
+    return bitCount(a ^ b);
+}
+
+int main() {
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:34]:
```suggestion
/**
 * @brief Main function
 * @returns 0 on exit
 */
int main() {
```

File: bit_manipulation/hamming_distance.cpp [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+ * have a set bit iff those bits differ in the two numbers.
+ * Hence, we return the number of such set bits.
+ *
+ * @author [Ravishankar Joshi](https://github.com/ravibitsgoa)
+ */
+
+#include <iostream>
+using std::cout, std::endl;
+
+unsigned int bitCount(unsigned int value) {
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:16]: It's better to use `uint64_t`, it's securer and better than `int` (check other parts of the code).

Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:16]: Missing function documentation (this is **necessary** and **required**).
See the [typical structure](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#typical-structure-of-a-program) of a program for more information. 🙂

File: bit_manipulation/hamming_distance.cpp [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+    while (value) {       // until all bits are zero
+        if (value & 1) {  // check lower bit
+            count++;
+        }
+        value >>= 1;  // shift bits, removing lower bit
+    }
+    return count;
+}
+
+unsigned int hamming_distance(int a, int b) {
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:27]: Missing function documentation.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Please enable [GitHub Actions](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions) in your repository of this fork in this link: https://github.com/ravibitsgoa/C-Plus-Plus/actions

File: bit_manipulation/hamming_distance.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+ * @details
+ * To find hamming distance between two integers, we take their xor, which will
+ * have a set bit iff those bits differ in the two numbers.
+ * Hence, we return the number of such set bits.
+ *
+ * @author [Ravishankar Joshi](https://github.com/ravibitsgoa)
+ */
+
+#include <cassert>
+#include <iostream>  /// for io operations
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:16]:
```suggestion
#include <cassert>   /// for assert
#include <iostream>  /// for io operations
```

File: bit_manipulation/hamming_distance.cpp [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:18]:
```suggestion
/**
 * @namespace bit_manipulation
 * @brief Bit Manipulation algorithms
 */
namespace bit_manipulation {
/**
```

File: bit_manipulation/hamming_distance.cpp [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+    return count;
+}
+
+/**
+ * This function returns the hamming distance between two integers.
+ * @param a the first number
+ * @param b the second number
+ * @returns the number of bits differing between the two integers.
+ */
+uint64_t hamming_distance(uint64_t a, uint64_t b) { return bitCount(a ^ b); }
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:40]:
```suggestion
uint64_t hamming_distance(uint64_t a, uint64_t b) { return bitCount(a ^ b); }
}  // namespace bit_manipulation
```

File: bit_manipulation/hamming_distance.cpp [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+/**
+ * @brief Function to the test hamming distance.
+ * @returns void
+ */
+static void test() {
+    assert(hamming_distance(11, 2) == 2);
+    assert(hamming_distance(2, 0) == 1);
+    assert(hamming_distance(11, 0) == 3);
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:49]:
```suggestion
    assert(bit_manipulation::hamming_distance(11, 2) == 2);
    assert(bit_manipulation::hamming_distance(2, 0) == 1);
    assert(bit_manipulation::hamming_distance(11, 0) == 3);
```

File: bit_manipulation/hamming_distance.cpp [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+}
+
+/**
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:57]:
```suggestion
    test();  // execute the tests
```

File: bit_manipulation/hamming_distance.cpp [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+    uint64_t a = 11;  // 1011 in binary
+    uint64_t b = 2;   // 0010 in binary
+
+    std::cout << "Hamming distance between " << a << " and " << b << " is "
+              << hamming_distance(a, b) << std::endl;
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:62]:
```suggestion
              << bit_manipulation::hamming_distance(a, b) << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

File: bit_manipulation/hamming_distance.cpp [hunk 11]
```diff
@@ -0,0 +1,69 @@
+/**
+ * @file hamming_distance.cpp
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:2]:
```suggestion
 * @file
```

**Review by Panquesito7 — APPROVED** [review]
Good work, thank you for your contribution! 😄 👍

**Comment by ravibitsgoa** [comment]
Hi @ayaankhan98 can you please review this PR and merge if found proper?

**Review by ayaankhan98 — COMMENTED** [review]
@ravibitsgoa we can have more generalized version of this algorithm, to consider strings also.

**Comment by ravibitsgoa** [comment]
> @ravibitsgoa we can have more generalized version of this algorithm, to consider strings also.

Yes, I can add a method for strings too. But the same method cannot accommodate strings right?

**Comment by ayaankhan98** [comment]
> > @ravibitsgoa we can have more generalized version of this algorithm, to consider strings also.
> 
> Yes, I can add a method for strings too. But the same method cannot accommodate strings right?
> Also, by strings, you mean hamming_distance("1111", "1101") == 1 right?

That's great, please do the implementation for strings too; A quick thought i have is you can use template specilization to achieve this for both strings and integers;
And yes you are right hamming distance for the above proposed strings is 1.

File: bit_manipulation/hamming_distance.cpp [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for io operations
+
+/**
+ * @namespace bit_manipulation
+ * @brief Bit Manipulation algorithms
+ */
+namespace bit_manipulation {
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:22]:
```suggestion
namespace bit_manipulation {
/**
 * @namespace hamming_distance
 * @brief Functions for [Hamming
 * distance](https://en.wikipedia.org/wiki/Hamming_distance) implementation
 */
namespace hamming_distance {
```

File: bit_manipulation/hamming_distance.cpp [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+    size_t n = a.size();
+    uint64_t count = 0;
+    for (size_t i = 0; i < n; i++) {
+        assert(a[i] == '0' || a[i] == '1');
+        assert(b[i] == '0' || b[i] == '1');
+        count += (b[i] != a[i]);
+    }
+    return count;
+}
+
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:64]:
```suggestion
}  // namespace hamming_distance
```

File: bit_manipulation/hamming_distance.cpp [hunk 14]
```diff
@@ -0,0 +1,92 @@
...
+}  // namespace bit_manipulation
+
+/**
+ * @brief Function to the test hamming distance.
+ * @returns void
+ */
+static void test() {
+    assert(bit_manipulation::hamming_distance(11, 2) == 2);
+    assert(bit_manipulation::hamming_distance(2, 0) == 1);
+    assert(bit_manipulation::hamming_distance(11, 0) == 3);
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:74]:
```suggestion
    assert(bit_manipulation::hamming_distance::hamming_distance(11, 2) == 2);
    assert(bit_manipulation::hamming_distance::hamming_distance(2, 0) == 1);
    assert(bit_manipulation::hamming_distance::hamming_distance(11, 0) == 3);
```

File: bit_manipulation/hamming_distance.cpp [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+
+    assert(bit_manipulation::hamming_distance("1101", "1111") == 1);
+    assert(bit_manipulation::hamming_distance("1111", "1111") == 0);
+    assert(bit_manipulation::hamming_distance("0000", "1111") == 4);
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:78]:
```suggestion
    assert(bit_manipulation::hamming_distance::hamming_distance("1101", "1111") == 1);
    assert(bit_manipulation::hamming_distance::hamming_distance("1111", "1111") == 0);
    assert(bit_manipulation::hamming_distance::hamming_distance("0000", "1111") == 4);
```

File: bit_manipulation/hamming_distance.cpp [hunk 16]
```diff
@@ -0,0 +1,92 @@
...
+ * @brief Main function
+ * @returns 0 on exit
+ */
+int main() {
+    test();           // execute the tests
+    uint64_t a = 11;  // 1011 in binary
+    uint64_t b = 2;   // 0010 in binary
+
+    std::cout << "Hamming distance between " << a << " and " << b << " is "
+              << bit_manipulation::hamming_distance(a, b) << std::endl;
```
Comment (Panquesito7) [inline bit_manipulation/hamming_distance.cpp:91]:
```suggestion
              << bit_manipulation::hamming_distance::hamming_distance(a, b) << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]

**Comment by ayaankhan98** [comment]
> > > @ravibitsgoa we can have more generalized version of this algorithm, to consider strings also.
> > 
> > 
> > Yes, I can add a method for strings too. But the same method cannot accommodate strings right?
> > Also, by strings, you mean hamming_distance("1111", "1101") == 1 right?
> 
> That's great, please do the implementation for strings too; A quick thought i have is you can use template specilization to achieve this for both strings and integers;
> And yes you are right hamming distance for the above proposed strings is 1.

This time i am mentioning this more explicitly by strings i mean all possible strings not just only binary strings.
consider alphabetical strings too.
```"cathy", "kathy" -> 1```

**Comment by ravibitsgoa** [comment]
> > > > @ravibitsgoa we can have more generalized version of this algorithm, to consider strings also.
> > > 
> > > 
> > > Yes, I can add a method for strings too. But the same method cannot accommodate strings right?
> > > Also, by strings, you mean hamming_distance("1111", "1101") == 1 right?
> > 
> > 
> > That's great, please do the implementation for strings too; A quick thought i have is you can use template specilization to achieve this for both strings and integers;
> > And yes you are right hamming distance for the above proposed strings is 1.
> 
> This time i am mentioning this more explicitly by strings i mean all possible strings not just only binary strings.
> consider alphabetical strings too.
> `"cathy", "kathy" -> 1`

Done.

**Review by Panquesito7 — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]

**Pull request merged**